extensions/file/src/main/java/org/mule/extension/file/internal/DirectoryListener.java

```diff
@@ -0,0 +1,487 @@
+        DataType dataType = DataTypeFactory.create(NullPayload.class);
+        MuleMessage<InputStream, ListenerFileAttributes> message;
+
+        if (attributes.getEventType() == FileEventType.DELETE)
+        {
+            attributes = new DeletedFileAttributes(path);
+        }
+        else if (!attributes.isDirectory())
+        {
+            dataType = fileSystem.getFileMessageDataType(DataTypeFactory.create(InputStream.class), attributes);
+            payload = new FileInputStream(path, new NullPathLock());
+        }
+
+        message = (MuleMessage) new DefaultMuleMessage(payload, dataType, attributes, muleContext);
+        return message;
+    }
+
+    @Override
+    public void stop()
+    {
+        stopRequested.set(true);
+        closeWatcherService();
+        shutdownExecutor();
+    }
+
+    private void shutdownExecutor()
+    {
+        if (executorService == null)
+        {
+            return;
+        }
+
+        executorService.shutdownNow();
+        try
+        {
+            if (!executorService.awaitTermination(muleContext.getConfiguration().getShutdownTimeout(), MILLISECONDS))
+            {
+                if (LOGGER.isWarnEnabled())
+                {
+                    LOGGER.warn("Could not properly terminate pending events for directory listener on flow " + flowConstruct.getName());
+                }
+            }
+        }
+        catch (InterruptedException e)
+        {
+            if (LOGGER.isWarnEnabled())
+            {
+                LOGGER.warn("Got interrupted while trying to terminate pending events for directory listener on flow " + flowConstruct.getName());
+            }
+        }
+    }
+
+    private void closeWatcherService()
+    {
+        if (watcher == null)
+        {
+            return;
+        }
+
+        try
+        {
+            watcher.close();
+        }
+        catch (IOException e)
+        {
+            if (LOGGER.isWarnEnabled())
+            {
+                LOGGER.warn("Found exception trying to close watcher service for directory listener on flow " + flowConstruct.getName(), e);
+            }
+        }
+
+        keyPaths.clear();
+    }
+
+    private void resetWatchKey(WatchKey key)
+    {
+        if (key.reset())
+        {
+            return;
+        }
+
+        Path path = keyPaths.remove(key);
+        if (path != null)
+        {
+            try
+            {
+                registerPath(path);
+            }
+            catch (IOException e)
+            {
+                if (LOGGER.isWarnEnabled())
+                {
+                    LOGGER.warn(format("Directory '%s' became unavailable and a new listener could not be established on it",
+                                       path.toString()));
+                }
+            }
+        }
+    }
+
+    private void createWatcherService() throws IOException
+    {
+        try
+        {
+            watcher = FileSystems.getDefault().newWatchService();
```

Comment:
no. The fileSystem you refer to is the connector's abstration of it. This is FileSystemS... a Java API class